dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,256 @@
+              | otherwise = Just (ln, lsl + 1 + lengthInBytes part1, part1 <> termAndNext)
+              where (lsl, l) = lineFromMap ln
+                    (part1, part2) = T.breakOn term l
+                    termAndNext = T.take (T.length term + 1) part2
+          fallbackInfoForText ln c = (ln, lsl + 1 + lengthInBytes pat, pat)
+              where (lsl, l) = lineFromMap ln
+                    pat = T.take c l
+
+getTagsFromExpr :: Expr Src a -> [(PosInFile, Text)]
+getTagsFromExpr = go (FP 0 0) []
+    where go lpos mts = \case
+              (Let b e) -> go lpos (mts <> parseBinding lpos b) e
+              (Annot e1 e2) -> go lpos (go lpos mts e1) e2
+              (Record mr) -> mts <> tagsFromDhallMap lpos mr
+              (RecordLit mr) -> mts <> tagsFromDhallMap lpos mr
+              (Union mmr) -> mts <> tagsFromDhallMapMaybe lpos mmr
+              (Note s e) -> go (srcToPosInFile s) mts e
+              _ -> mts
+          tagsFromDhallMap lpos = foldMapWithKey (tagsFromDhallMapElement lpos)
+          tagsFromDhallMapMaybe lpos = foldMapWithKey (\k -> \case
+              Just e -> tagsFromDhallMapElement lpos k e
+              _ -> [(lpos, k)])
+          tagsFromDhallMapElement lpos k e = go pos [(pos, k)] e
+              where pos = firstPosFromExpr lpos e
+          parseBinding :: PosInFile -> Binding Src a -> [(PosInFile, Text)]
+          parseBinding lpos b = go p2 [(p0, variable b)] (value b)
+              where p0 = posFromBinding (bindingSrc0 b) lpos
+                    p1 = posFromBinding (bindingSrc1 b) p0
+                    p2 = posFromBinding (bindingSrc2 b) p1
+          posFromBinding src startPos = maybe startPos srcToPosInFile src
+
+srcToPosInFile :: Src -> PosInFile
+srcToPosInFile s = FP line column
+    where ssp = srcStart s
+          line = unPos . sourceLine $ ssp
+          column = unPos . sourceColumn $ ssp
+
+firstPosFromExpr :: PosInFile -> Expr Src a -> PosInFile
+firstPosFromExpr lpos = \case
+    (Note s _) -> srcToPosInFile s
+    _ -> lpos
+
+showTags :: Tags -> Text
+showTags (Tags ts) = T.concat . map (uncurry showFileTags) . M.toList $ ts
+
+showFileTags :: FilePath -> [(PosInFile, Tag)] -> T.Text
+showFileTags f ts = "\x0c\n" <> T.pack f <> "," <> (showInt . T.length) cs <> "\n" <> cs
+    where cs = T.concat . map (uncurry showPosTag) $ ts
+
+showPosTag :: PosInFile -> Tag -> Text
+showPosTag fp tag = def <>"\x7f" <> name <> "\x01" <> showInt line <>
+                    "," <> showInt offset <> "\n"
+    where line = fpLine fp
+          offset = fpOffset fp
+          def = tagPattern tag
+          name = tagName tag
+
+showInt :: Int -> Text
+showInt = T.pack . show
+
+{-| Generate list of files for given Input
+-}
+inputToFiles
+    :: Bool
+    -- ^ In case true, function will get files from  symbolic links
+    -> [String]
+    -- ^ List of suffixes. In case it contains empty string (""), all files will be returned.
```

Comment:
We probably should not use a sentinel value (the empty string) to specify returning all files.

You can use `Maybe [String]` where `Nothing` means to return all suffixes.

If you want to really get fancy (at the cost of another dependency) you could use the `Optional` type from my `optional-args` package here: http://hackage.haskell.org/package/optional-args-1.0.2/docs/Data-Optional.html .  However, using `Maybe` is probably good enough here

Reply:
I changed it to Maybe.